psycop/common/model_training_v2/trainer/selective_cross_validator_trainer.py

```diff
@@ -0,0 +1,111 @@
+from dataclasses import dataclass
+
+import pandas as pd
+import polars as pl
+from sklearn.model_selection import StratifiedGroupKFold
+
+from psycop.common.model_training_v2.config.baseline_registry import BaselineRegistry
+from psycop.common.model_training_v2.trainer.base_dataloader import BaselineDataLoader
+from psycop.common.model_training_v2.trainer.base_trainer import BaselineTrainer, TrainingResult
+from psycop.common.model_training_v2.trainer.preprocessing.pipeline import PreprocessingPipeline
+from psycop.common.model_training_v2.trainer.task.base_metric import BaselineMetric
+from psycop.common.model_training_v2.trainer.task.base_task import BaselineTask
+
+
+@BaselineRegistry.trainers.register("selective_crossval_trainer")
+@dataclass
+class SelectiveCrossValidatorTrainer(BaselineTrainer):
+    uuid_col_name: str
+    outcome_col_name: str
+    training_data: BaselineDataLoader
+    additional_data: BaselineDataLoader
+    preprocessing_pipeline: PreprocessingPipeline
+    task: BaselineTask
+    metric: BaselineMetric
+    n_splits: int = 5
+    group_col_name: str = "dw_ek_borger"
+    """A cross-validator trainer that trains on a combination of `training_data`
+    and `additional_data` and evaluates on the `training_data` only. This can be
+    useful for training with synthetic data and only evluating on the real data
+    or for training with prevalent cases and evaluating on incident cases only."""
+
+    def train(self) -> TrainingResult:
+        # assign a dummy variable "_do_eval" do indicate whether the data point should be used for evaluation
+        training_data_preprocessed = self.preprocessing_pipeline.apply(
+            data=self.training_data.load()
+        ).assign(_do_eval=1)
+
+        additional_data = self.additional_data.load().collect().to_pandas().assign(_do_eval=0)
+        training_data_preprocessed = pd.concat(
+            [training_data_preprocessed, additional_data], axis=0
+        ).reset_index(drop=True)
+
+        training_data_preprocessed["dw_ek_borger"] = training_data_preprocessed[
+            "dw_ek_borger"
+        ].astype(str)
+
+        X = training_data_preprocessed.drop([self.outcome_col_name, self.uuid_col_name], axis=1)
+        y = pd.DataFrame(
+            training_data_preprocessed[self.outcome_col_name], columns=[self.outcome_col_name]
+        )
+        self.logger.info(f"\tOutcome: {list(y.columns)}")
+
+        folds = StratifiedGroupKFold(n_splits=self.n_splits).split(
+            X=X, y=y, groups=training_data_preprocessed[self.group_col_name]
```

Comment:
Burde man også tage højde for _do_eval når man splitters foldsne for at forsgøge at have lige meget additional_data i hvert fold? Ellers kan hvert test split måske variere en del i størrelse når additional_data fjernes, hvilket måske kan være lidt messy at rapportere?

Reply:
Helt fair bekymring! Jeg ved ikke hvor meget det betyder i praksis (nok ret udfaldsspecifikt), men `StratifiedGroupKFold` kan desværre kun tage én variabel. Kræver nok at man lige tester - kan evt. logge størrelsen på eval splitsene til loggeren

Reply:
Ja, måske fint bare at logge split-størrelserne? Så kan man løbende nemt overvåge effekten når man køre et eksperiment, hvis det er.

Reply:
Yes, det implementerer jeg før merge - god pointe!